File: ios/WaveformBarsView.swift
import UIKit

/// Custom UIView that draws the audio waveform as a row of vertical rounded-rect
/// bars with a "played" / "unplayed" two-tone fill.
///
/// Drawing model:
///   1. Build a single cached `UIBezierPath` containing every bar's rounded-rect
///      whenever the amplitudes / size / bar geometry change.
///   2. On every `draw(_:)`:
///        a. Fill the cached path in `unplayedBarColor`.
///        b. Save state -> clip to `[0, progressX, bounds.height]` ->
///           fill the same cached path in `playedBarColor` -> restore state.
///      The bar straddling the playhead is partial-highlighted naturally because
///      the rounded-rect geometry is identical in both passes; only the right
///      edge is cropped by the clip.
///
/// Touch handling is **immediate** (no slop, no long-press timeout): scrubbing
/// starts on `touchesBegan` and the host view receives the proportional position
/// via the `onScrubBegan/Moved/Ended` callbacks.
final class WaveformBarsView: UIView {

    // MARK: - Visual configuration (set by AudioWaveformViewImpl)

    var amplitudes: [CGFloat] = [] {
        didSet { setTargetAmplitudes(amplitudes) }
    }

    var playedBarColor: UIColor = .white {
        didSet { setNeedsDisplay() }
    }

    var unplayedBarColor: UIColor = UIColor.white.withAlphaComponent(0.5) {
        didSet { setNeedsDisplay() }
    }

    var barWidth: CGFloat = 3 {
        didSet { invalidatePath() }
    }

    var barGap: CGFloat = 2 {
        didSet { invalidatePath() }
    }

    /// `< 0` means "auto" = barWidth / 2.
    var barRadius: CGFloat = -1 {
        didSet { invalidatePath() }
    }

    /// `<= 0` means "auto from view width".
    var barCountOverride: Int = 0 {
        didSet { invalidatePath() }
    }

    /// Fraction of the waveform that has been played, in [0, 1].
    /// Updated at ~30 Hz while playing. The cached bar path makes the
    /// per-frame redraw essentially free, so we always invalidate.
    var progressFraction: CGFloat = 0 {
        didSet {
            if oldValue != progressFraction {
                setNeedsDisplay()
            }
        }
    }

    // MARK: - Touch / scrub callbacks

    /// All callbacks pass a fraction in [0, 1] (the touch's x divided by
    /// the view's width).
    var onScrubBegan: ((CGFloat) -> Void)?
    var onScrubMoved: ((CGFloat) -> Void)?
    var onScrubEnded: ((CGFloat, _ cancelled: Bool) -> Void)?

    // MARK: - Private state

    private var cachedPath: UIBezierPath?
    private var cachedSize: CGSize = .zero

    /// What's actually drawn this frame. During an amplitudes update, these
    /// values smoothly interpolate between `startAmps` and `targetAmps`.
    private var displayedAmps: [CGFloat] = []
    private var startAmps: [CGFloat] = []
    private var targetAmps: [CGFloat] = []
    private var amplitudeAnimationStart: CFTimeInterval = 0
    private var amplitudeDisplayLink: CADisplayLink?
    /// How long each new partial-amplitudes update animates for.
    private static let amplitudeAnimationDuration: CFTimeInterval = 0.2

    // MARK: - Init

    private lazy var scrubRecognizer: UILongPressGestureRecognizer = {
        // `minimumPressDuration = 0` makes the gesture activate immediately on
        // touch-down. `allowableMovement = .greatestFiniteMagnitude` keeps it
        // alive through any drag distance.
        //
        // Long-press is the canonical iOS pattern for "claim the touch
        // immediately so a parent UIScrollView can't cancel it" — once a
        // long-press recognizer has begun, UIScrollView's pan can no longer
        // hijack the touch sequence (which is what was causing scrubbing to
        // fail when the bars view was nested inside a ScrollView).
        let r = UILongPressGestureRecognizer(target: self, action: #selector(handleScrubGesture(_:)))
        r.minimumPressDuration = 0
        r.allowableMovement = .greatestFiniteMagnitude
        r.cancelsTouchesInView = false
        return r
    }()

    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        commonInit()
    }

    private func commonInit() {
        isOpaque = false
        backgroundColor = .clear
        contentMode = .redraw
        isUserInteractionEnabled = true
        addGestureRecognizer(scrubRecognizer)
    }

    deinit {
        amplitudeDisplayLink?.invalidate()
        amplitudeDisplayLink = nil
    }

    // MARK: - Layout / path caching

    override func layoutSubviews() {
        super.layoutSubviews()
        if cachedSize != bounds.size {
            invalidatePath()
        }
    }

    private func invalidatePath() {
        cachedPath = nil
        cachedSize = .zero
        setNeedsDisplay()
    }

    private func ensureCachedPath() -> UIBezierPath? {
        if let path = cachedPath, cachedSize == bounds.size {
            return path
        }
        let path = buildBarPath()
        cachedPath = path
        cachedSize = bounds.size
        return path
    }

    /// Uniform amplitude used to render the "loading" skeleton when no real
    /// amplitudes have been decoded yet. Small enough to read clearly as a
    /// placeholder, big enough to be obviously visible.
    private static let placeholderAmplitude: CGFloat = 0.2

    /// Build a composite `UIBezierPath` of every bar's rounded-rect. This is
    /// rebuilt whenever amplitudes / size / bar geometry change (and on every
    /// frame while an amplitude animation is running), so the hot 30 Hz
    /// playback redraw can reuse the cached path.
    ///
    /// When no amplitudes have ever been delivered (decoder hasn't produced
    /// the first partial yet) we render a uniform low-amplitude skeleton so
    /// the user sees the bar pattern immediately instead of an empty card.
    private func buildBarPath() -> UIBezierPath? {
        let totalWidth = bounds.width
        let totalHeight = bounds.height
        guard totalWidth > 0, totalHeight > 0 else { return nil }

        let step = barWidth + barGap
        guard step > 0 else { return nil }

        let autoCount = Int(floor(totalWidth / step))
        let barCount = barCountOverride > 0
            ? min(barCountOverride, autoCount)
            : autoCount
        guard barCount > 0 else { return nil }

        let verticalPadding = barWidth * 1.5
        let drawableHeight = totalHeight - verticalPadding * 2
        guard drawableHeight > 0 else { return nil }
        let minBarHeight = barWidth
        let radius = barRadius < 0 ? barWidth / 2 : barRadius
        // `displayedAmps` already substitutes `placeholderAmplitude` for any
        // not-yet-decoded bar in `setTargetAmplitudes`, so trailing bars stay
        // at skeleton height instead of dropping to `minBarHeight`.
        let usePlaceholder = displayedAmps.isEmpty

        let path = UIBezierPath()
        for i in 0..<barCount {
            let amp: CGFloat
            if usePlaceholder {
                amp = Self.placeholderAmplitude
            } else {
                let ampIndex = i * displayedAmps.count / barCount
                amp = displayedAmps[min(max(ampIndex, 0), displayedAmps.count - 1)]
            }
            let barHeight = max(minBarHeight, amp * drawableHeight)
            let x = CGFloat(i) * step
            let y = verticalPadding + (drawableHeight - barHeight) / 2.0
            let rect = CGRect(x: x, y: y, width: barWidth, height: barHeight)
            path.append(UIBezierPath(roundedRect: rect, cornerRadius: radius))
        }
        return path
    }

    // MARK: - Amplitude animation

    /// Update the animation targets when new amplitudes arrive. Each new
    /// partial smoothly grows from the currently-displayed values to the
    /// new target over `amplitudeAnimationDuration` (ease-out cubic). Bars
    /// expand symmetrically from the centre because the bar geometry is
    /// already centre-aligned (`y = verticalPadding + (drawableHeight - h)/2`).
    private func setTargetAmplitudes(_ newAmps: [CGFloat]) {
        if newAmps.isEmpty {
            stopAmplitudeAnimation()
            displayedAmps = []
            startAmps = []
            targetAmps = []
            invalidatePath()
            return
        }

        // Treat zero amplitude as "not yet decoded" — keep those bars at the
        // skeleton placeholder height instead of letting them snap to
        // `minBarHeight` between partials.
        let processed: [CGFloat] = newAmps.map { $0 > 0 ? $0 : Self.placeholderAmplitude }

        if displayedAmps.count != processed.count {
            // First non-empty payload (or barCount changed). Seed the
            // current values with the placeholder skeleton so the animation
            // grows from skeleton -> real shape.
            displayedAmps = [CGFloat](repeating: Self.placeholderAmplitude, count: processed.count)
        }

        startAmps = displayedAmps
        targetAmps = processed
        amplitudeAnimationStart = CACurrentMediaTime()
        startAmplitudeAnimation()
    }

    private func startAmplitudeAnimation() {
        if amplitudeDisplayLink != nil { return }
        let link = CADisplayLink(target: self, selector: #selector(handleAmplitudeTick))
        link.add(to: .main, forMode: .common)
        amplitudeDisplayLink = link
    }

    private func stopAmplitudeAnimation() {
        amplitudeDisplayLink?.invalidate()
        amplitudeDisplayLink = nil
    }

    @objc private func handleAmplitudeTick() {
        let elapsed = CACurrentMediaTime() - amplitudeAnimationStart
        let t = max(0, min(1, elapsed / Self.amplitudeAnimationDuration))
        let eased = Self.easeOutCubic(CGFloat(t))
        let count = min(displayedAmps.count, min(startAmps.count, targetAmps.count))
        for i in 0..<count {
            displayedAmps[i] = startAmps[i] + (targetAmps[i] - startAmps[i]) * eased
        }
        invalidatePath()
        if t >= 1 {
            stopAmplitudeAnimation()
        }
    }

    private static func easeOutCubic(_ t: CGFloat) -> CGFloat {
        let inv = 1 - t
        return 1 - inv * inv * inv
    }

    // MARK: - Drawing

    override func draw(_ rect: CGRect) {
        guard let path = ensureCachedPath() else { return }
        guard let ctx = UIGraphicsGetCurrentContext() else { return }

        // Pass 1: fill all bars in the unplayed color.
        unplayedBarColor.setFill()
        path.fill()

        // Pass 2: clip to [0, progressX] and fill the same path in the played color.
        let clamped = max(0, min(1, progressFraction))
        let progressX = clamped * bounds.width
        guard progressX > 0 else { return }

        ctx.saveGState()
        ctx.clip(to: CGRect(x: 0, y: 0, width: progressX, height: bounds.height))
        playedBarColor.setFill()
        path.fill()
        ctx.restoreGState()
    }

    // MARK: - Touch handling (immediate scrub via long-press gesture)

    @objc private func handleScrubGesture(_ recognizer: UILongPressGestureRecognizer) {
        let location = recognizer.location(in: self)
        let fraction = clampFraction(location.x / max(1, bounds.width))
        switch recognizer.state {
        case .began:
            progressFraction = fraction
            setNeedsDisplay()
            onScrubBegan?(fraction)
        case .changed:
            progressFraction = fraction
            setNeedsDisplay()
            onScrubMoved?(fraction)
        case .ended:
            progressFraction = fraction
            setNeedsDisplay()
            onScrubEnded?(fraction, false)
        case .cancelled, .failed:
            onScrubEnded?(progressFraction, true)
        default:
            break
        }
    }

    private func clampFraction(_ value: CGFloat) -> CGFloat {
        if value.isNaN { return 0 }
        return max(0, min(1, value))
    }
}

File: android/src/main/java/com/audiowaveform/WaveformBarsView.kt
package com.audiowaveform

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.animation.DecelerateInterpolator

/**
 * Renders the audio waveform as a row of vertical rounded-rect bars with a
 * "played" / "unplayed" two-tone fill, and handles immediate touch-and-drag
 * scrubbing.
 *
 * Drawing model:
 *   1. Build a single cached `Path` containing every bar's rounded-rect when
 *      amplitudes / size / bar geometry change.
 *   2. On every `onDraw`:
 *        a. `canvas.drawPath(cachedPath, unplayedPaint)`.
 *        b. `canvas.save()` -> `clipRect(0, 0, progressX, height)` ->
 *           `canvas.drawPath(cachedPath, playedPaint)` -> `canvas.restore()`.
 *
 * Touch handling is **immediate** — scrub starts at `ACTION_DOWN`, with no
 * slop or long-press delay. The host view receives the proportional position
 * via the `onScrubBegan/Moved/Ended` callbacks.
 */
class WaveformBarsView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : View(context, attrs, defStyleAttr) {

    // region Visual configuration ------------------------------------------------

    var amplitudes: FloatArray = FloatArray(0)
        private set

    /**
     * Update the bar amplitudes. Each call animates the existing displayed
     * heights toward the new target over [AMPLITUDE_ANIMATION_DURATION_MS]
     * (ease-out). Bars expand symmetrically from the centre because the bar
     * geometry is centre-aligned in [buildBarPath].
     */
    fun setAmplitudes(values: FloatArray) {
        amplitudes = values
        setTargetAmplitudes(values)
    }

    var playedBarColor: Int
        get() = playedPaint.color
        set(value) {
            playedPaint.color = value
            invalidate()
        }

    var unplayedBarColor: Int
        get() = unplayedPaint.color
        set(value) {
            unplayedPaint.color = value
            invalidate()
        }

    var barWidthPx: Float = 3f * resources.displayMetrics.density
        set(value) {
            field = value
            invalidatePath()
        }

    var barGapPx: Float = 2f * resources.displayMetrics.density
        set(value) {
            field = value
            invalidatePath()
        }

    /** `< 0` means "auto" = barWidth / 2. */
    var barRadiusPx: Float = -1f
        set(value) {
            field = value
            invalidatePath()
        }

    /** `<= 0` means "auto from view width". */
    var barCountOverride: Int = 0
        set(value) {
            field = value
            invalidatePath()
        }

    /** Fraction of the waveform that has been played, in [0, 1]. */
    var progressFraction: Float = 0f
        set(value) {
            val clamped = value.coerceIn(0f, 1f)
            if (clamped != field) {
                field = clamped
                invalidate()
            }
        }

    // endregion

    // region Touch / scrub callbacks ---------------------------------------------

    /** All callbacks pass a fraction in [0, 1] (touch x / view width). */
    var onScrubBegan: ((Float) -> Unit)? = null
    var onScrubMoved: ((Float) -> Unit)? = null
    var onScrubEnded: ((Float, Boolean /* cancelled */) -> Unit)? = null

    // endregion

    // region Private state -------------------------------------------------------

    private val playedPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.WHITE
        style = Paint.Style.FILL
    }
    private val unplayedPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
        color = Color.argb(128, 255, 255, 255)
        style = Paint.Style.FILL
    }

    private var cachedPath: Path? = null
    private var cachedWidth: Int = 0
    private var cachedHeight: Int = 0
    private val tmpRect = RectF()

    /**
     * Values actually drawn this frame. While an amplitudes update is
     * animating, these are linearly interpolated between [startAmps] and
     * [targetAmps] by [amplitudeAnimator].
     */
    private var displayedAmps: FloatArray = FloatArray(0)
    private var startAmps: FloatArray = FloatArray(0)
    private var targetAmps: FloatArray = FloatArray(0)

    private val amplitudeAnimator: ValueAnimator =
        ValueAnimator.ofFloat(0f, 1f).apply {
            duration = AMPLITUDE_ANIMATION_DURATION_MS
            interpolator = DecelerateInterpolator(1.5f)
            addUpdateListener { va ->
                val t = va.animatedValue as Float
                val count = minOf(displayedAmps.size, startAmps.size, targetAmps.size)
                for (i in 0 until count) {
                    displayedAmps[i] = startAmps[i] + (targetAmps[i] - startAmps[i]) * t
                }
                invalidatePath()
            }
        }

    // endregion

    init {
        isClickable = true
        isFocusable = true
    }

    // region Layout / path caching -----------------------------------------------

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)
        if (w != cachedWidth || h != cachedHeight) {
            invalidatePath()
        }
    }

    private fun invalidatePath() {
        cachedPath = null
        cachedWidth = 0
        cachedHeight = 0
        invalidate()
    }

    private fun ensureCachedPath(): Path? {
        val current = cachedPath
        if (current != null && cachedWidth == width && cachedHeight == height) {
            return current
        }
        val built = buildBarPath() ?: return null
        cachedPath = built
        cachedWidth = width
        cachedHeight = height
        return built
    }

    /**
     * Build the cached path. When [displayedAmps] is empty (decoding hasn't
     * produced anything yet) we render a uniform low-amplitude skeleton so
     * the user sees the bar pattern immediately instead of an empty card.
     *
     * `displayedAmps` already substitutes [PLACEHOLDER_AMPLITUDE] for any
     * not-yet-decoded bar in [setTargetAmplitudes], so trailing bars stay at
     * skeleton height instead of dropping to `minBarHeight` between partials.
     */
    private fun buildBarPath(): Path? {
        val totalWidth = width.toFloat()
        val totalHeight = height.toFloat()
        if (totalWidth <= 0 || totalHeight <= 0) return null

        val step = barWidthPx + barGapPx
        if (step <= 0f) return null

        val autoCount = (totalWidth / step).toInt()
        val barCount = if (barCountOverride > 0) {
            kotlin.math.min(barCountOverride, autoCount)
        } else {
            autoCount
        }
        if (barCount <= 0) return null

        val verticalPadding = barWidthPx * 1.5f
        val drawableHeight = totalHeight - verticalPadding * 2f
        if (drawableHeight <= 0f) return null
        val minBarHeight = barWidthPx
        val radius = if (barRadiusPx < 0f) barWidthPx / 2f else barRadiusPx
        val usePlaceholder = displayedAmps.isEmpty()

        val path = Path()
        for (i in 0 until barCount) {
            val amp = if (usePlaceholder) {
                PLACEHOLDER_AMPLITUDE
            } else {
                val ampIndex = (i * displayedAmps.size / barCount)
                    .coerceIn(0, displayedAmps.size - 1)
                displayedAmps[ampIndex].coerceIn(0f, 1f)
            }
            val barHeight = (amp * drawableHeight).coerceAtLeast(minBarHeight)
            val x = i * step
            val y = verticalPadding + (drawableHeight - barHeight) / 2f
            tmpRect.set(x, y, x + barWidthPx, y + barHeight)
            path.addRoundRect(tmpRect, radius, radius, Path.Direction.CW)
        }
        return path
    }

    /**
     * Update the animation targets when new amplitudes arrive. Each call
     * captures the current [displayedAmps] as the start, sets the new
     * `targetAmps`, and (re)starts the animator. Re-entrant — a new payload
     * mid-animation cancels the previous run and animates from the
     * mid-frame state to the new target.
     */
    private fun setTargetAmplitudes(values: FloatArray) {
        if (values.isEmpty()) {
            amplitudeAnimator.cancel()
            displayedAmps = FloatArray(0)
            startAmps = FloatArray(0)
            targetAmps = FloatArray(0)
            invalidatePath()
            return
        }

        // Treat zero amplitude as "not yet decoded" — keep those bars at the
        // skeleton placeholder height instead of letting them snap to
        // `minBarHeight` between partials.
        val processed = FloatArray(values.size) { i ->
            if (values[i] > 0f) values[i] else PLACEHOLDER_AMPLITUDE
        }

        if (displayedAmps.size != processed.size) {
            // First non-empty payload (or barCount changed). Seed the
            // current values with the placeholder skeleton so the animation
            // grows from skeleton -> real shape.
            displayedAmps = FloatArray(processed.size) { PLACEHOLDER_AMPLITUDE }
        }

        startAmps = displayedAmps.copyOf()
        targetAmps = processed
        amplitudeAnimator.cancel()
        amplitudeAnimator.start()
    }

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        amplitudeAnimator.cancel()
    }

    companion object {
        /**
         * Uniform amplitude used to render the "loading" skeleton when no
         * real amplitudes have been decoded yet.
         */
        private const val PLACEHOLDER_AMPLITUDE = 0.2f

        /** Duration (ms) of each amplitude-update animation. */
        private const val AMPLITUDE_ANIMATION_DURATION_MS = 200L
    }

    // endregion

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        val path = ensureCachedPath() ?: return

        // Pass 1: all bars in the unplayed color.
        canvas.drawPath(path, unplayedPaint)

        // Pass 2: clip to [0, progressX] and re-fill in the played color.
        val progressX = (progressFraction.coerceIn(0f, 1f) * width.toFloat())
        if (progressX <= 0f) return
        val saveCount = canvas.save()
        canvas.clipRect(0f, 0f, progressX, height.toFloat())
        canvas.drawPath(path, playedPaint)
        canvas.restoreToCount(saveCount)
    }

    // region Touch handling — immediate scrub ------------------------------------

    @Suppress("ClickableViewAccessibility")
    override fun onTouchEvent(event: MotionEvent): Boolean {
        val w = width
        if (w <= 0) return super.onTouchEvent(event)

        when (event.actionMasked) {
            MotionEvent.ACTION_DOWN -> {
                // Don't let a parent ScrollView steal the drag.
                parent?.requestDisallowInterceptTouchEvent(true)
                val fraction = (event.x / w.toFloat()).coerceIn(0f, 1f)
                progressFraction = fraction
                invalidate()
                onScrubBegan?.invoke(fraction)
                return true
            }
            MotionEvent.ACTION_MOVE -> {
                val fraction = (event.x / w.toFloat()).coerceIn(0f, 1f)
                progressFraction = fraction
                invalidate()
                onScrubMoved?.invoke(fraction)
                return true
            }
            MotionEvent.ACTION_UP -> {
                val fraction = (event.x / w.toFloat()).coerceIn(0f, 1f)
                progressFraction = fraction
                invalidate()
                onScrubEnded?.invoke(fraction, false)
                parent?.requestDisallowInterceptTouchEvent(false)
                performClick()
                return true
            }
            MotionEvent.ACTION_CANCEL -> {
                onScrubEnded?.invoke(progressFraction, true)
                parent?.requestDisallowInterceptTouchEvent(false)
                return true
            }
        }
        return super.onTouchEvent(event)
    }

    override fun performClick(): Boolean {
        super.performClick()
        return true
    }

    // endregion
}
